library/ssl_tls13_client.c

```diff
@@ -1141,6 +1141,11 @@ int mbedtls_ssl_tls13_write_client_hello_exts(mbedtls_ssl_context *ssl,
 
     *out_len = 0;
 
+    ret = mbedtls_ssl_tls13_crypto_init(ssl);
```

Comment:
I think this would more naturally fit in `ssl_prepare_client_hello()` in `ssl_client.c` where we know for the first time if we are going to propose TLS 1.3 to the server or not: after lines 771 to 783 if `ssl->tls_version == MBEDTLS_SSL_VERSION_TLS1_3`.

After the TLS 1.2 specific code (from line 815 to 844), instead of:
```
#if defined(MBEDTLS_SSL_TLS1_3_COMPATIBILITY_MODE)                              
    if (ssl->tls_version == MBEDTLS_SSL_VERSION_TLS1_3) {                       
        /*                                                                      
         * Create a legacy session identifier for the purpose of middlebox      
         * compatibility only if one has not been created already, which is     
         * the case if we are here for the TLS 1.3 second ClientHello.          
         *                                                                      
         * Versions of TLS before TLS 1.3 supported a "session resumption"      
         * feature which has been merged with pre-shared keys in TLS 1.3        
         * version. A client which has a cached session ID set by a pre-TLS 1.3 
         * server SHOULD set this field to that value. In compatibility mode,   
         * this field MUST be non-empty, so a client not offering a pre-TLS 1.3 
         * session MUST generate a new 32-byte value. This value need not be    
         * random but SHOULD be unpredictable to avoid implementations fixating 
         * on a specific value (also known as ossification). Otherwise, it MUST 
         * be set as a zero-length vector ( i.e., a zero-valued single byte     
         * length field ).                                                      
         */                                                                     
        session_id_len = 32;                                                    
    }                                                                           
#endif /* MBEDTLS_SSL_TLS1_3_COMPATIBILITY_MODE */
```
we could have something like (not tested):
```
#if defined(MBEDTLS_SSL_PROTO_TLS1_3)                              
    if (ssl->tls_version == MBEDTLS_SSL_VERSION_TLS1_3) { 
        ret = mbedtls_ssl_tls13_crypto_init(ssl);
        if (ret != 0) {
            return ret;
        }
            
#if defined(MBEDTLS_SSL_TLS1_3_COMPATIBILITY_MODE)
        /*                                                                      
         * Create a legacy session identifier for the purpose of middlebox      
         * compatibility only if one has not been created already, which is     
         * the case if we are here for the TLS 1.3 second ClientHello.          
         *                                                                      
         * Versions of TLS before TLS 1.3 supported a "session resumption"      
         * feature which has been merged with pre-shared keys in TLS 1.3        
         * version. A client which has a cached session ID set by a pre-TLS 1.3 
         * server SHOULD set this field to that value. In compatibility mode,   
         * this field MUST be non-empty, so a client not offering a pre-TLS 1.3 
         * session MUST generate a new 32-byte value. This value need not be    
         * random but SHOULD be unpredictable to avoid implementations fixating 
         * on a specific value (also known as ossification). Otherwise, it MUST 
         * be set as a zero-length vector ( i.e., a zero-valued single byte     
         * length field ).                                                      
         */                                                                     
        session_id_len = 32;
#endif                                                    
    }                                                                           
#endif /* MBEDTLS_SSL_PROTO_TLS1_3 */
```
Wdyt?

Reply:
So a little earlier, that should work. But it means changing a non-13 file, which I'd prefer to avoid. Also it's buried right in the middle of a function, so a failure there (admittedly unlikely) could leave the context in a worse state. On balance I prefer the place where it is now. And I'd rather avoid one more round of rework given the urgency.

Reply:
Okay, fair enough, let's go with what we have.